src/nupic/algorithms/SDRClassifier.cpp

```diff
@@ -0,0 +1,745 @@
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+#include <cmath>
+#include <deque>
+#include <iostream>
+#include <limits>
+#include <map>
+#include <string>
+#include <sstream>
+#include <vector>
+#include <stdio.h>
+
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include <kj/std/iostream.h>
+
+#include <nupic/algorithms/ClassifierResult.hpp>
+#include <nupic/algorithms/SDRClassifier.hpp>
+#include <nupic/proto/SdrClassifier.capnp.h>
+#include <nupic/math/ArrayAlgo.hpp>
+#include <nupic/utils/Log.hpp>
+
+using namespace std;
+
+namespace nupic
+{
+  namespace algorithms 
+  {
+    namespace sdr_classifier
+    {
+
+      SDRClassifier::SDRClassifier(
+          const vector<UInt>& steps, Real64 alpha, Real64 actValueAlpha,
+          UInt verbosity) : alpha_(alpha), actValueAlpha_(actValueAlpha),
+          learnIteration_(0), recordNumMinusLearnIteration_(0), 
+          maxInputIdx_(0), maxBucketIdx_(0), version_(Version), 
+          verbosity_(verbosity)
+      {
+        for (const auto& step : steps)
+        {
+          steps_.push_back(step);
+        }
+        recordNumMinusLearnIterationSet_ = false;
+        maxSteps_ = 0;
+        for (auto& elem : steps_)
+        {
+          UInt current = elem + 1;
+          if (current > maxSteps_)
+          {
+            maxSteps_ = current;
+          }
+        }
+        actualValues_.push_back(0.0);
+        actualValuesSet_.push_back(false);
+
+        // TODO: insert maxBucketIdx / maxInputIdx hint as parameter?
+        // There can be great overhead reallocating the array every time a new
+        // input is seen, especially if we start at (0, 0). The client will
+        // usually know what is the final maxInputIdx (typically the number
+        // of columns?), and we can have heuristics using the encoder's 
+        // settings to get an good approximate of the maxBucketIdx, thus having
+        // to reallocate this matrix only a few times, even never if we use
+        // lower bounds
+        for (const auto& step : steps_)
+        {
+          Matrix weights = Matrix(maxInputIdx_ + 1, maxBucketIdx_ + 1);
+          weightMatrix_.insert(pair<UInt, Matrix>(step, weights));
+        }
+      }
+
+      SDRClassifier::~SDRClassifier()
+      {
+      }
+
+      void SDRClassifier::compute(
+        UInt recordNum, const vector<UInt>& patternNZ, UInt bucketIdx,
+        Real64 actValue, bool category, bool learn, bool infer,
+        ClassifierResult* result)
+      {
+        // save the offset between recordNum and learnIteration_ if this
+        // was not set (first call to compute)
+        if (!recordNumMinusLearnIterationSet_)
+        {
+          recordNumMinusLearnIteration_ = recordNum - learnIteration_;
```

Comment:
@mrcslws you are right, simply storing recordNumHistory would be enough. But changing this would have repercussion on the proto, thus on the Python implementation, including the Region, and raise consistency issues with the CLAClassifier in those regions. I think I should leave that as is for now.